src/book.rs

```diff
@@ -152,162 +197,114 @@ impl Book {
-            {
-                if orders_queue.len() > 1 {
-                    continue;
-                } else {
-                    break;
+                /* if we haven't crossed the spread, we're not going to match */
+                if opposing_top.is_none() || opposing_top.unwrap() > order.price
+                {
+                    return self.add_order(order);
                 }
-            }
 
-            // iterate over orders at the current "best" price
-            while let Some(mut matching_order) = orders_queue.pop_front() {
-                // compute new amounts for our orders using temp variables
-                let match_to_submit = matching_order.clone();
-                let mut matching_amount = matching_order.amount;
-                matching_order.amount =
-                    matching_amount.saturating_sub(order_amount);
-                order_amount = order_amount.saturating_sub(matching_amount);
-                matching_amount = matching_order.amount;
-
-                // forward to the contracts
-                info!("Forwarding ({},{})", order, match_to_submit);
-
-                /* push to contract */
-                match rpc::send_matched_orders(
-                    order.clone(),
-                    match_to_submit,
-                    executioner_address.clone(),
-                )
-                .await
-                {
-                    Ok(t) => {
-                        info!("See https://etherscan.io/tx/{}", t);
+                for (price, asks) in self.asks.iter_mut() {
+                    /* if we've run out of viable prices or we're done, halt */
+                    if done || *price > order.price {
+                        break;
                     }
-                    Err(e) => {
-                        warn!("Failed to forward matched orders! {}", e);
-                        // todo: how should we handle this when no executioner is running
-                        // eg in tests?
-                        //return Err(BookError::Web3Error);
-                    }
-                };
-
-                info!("Executed {}", matching_order);
 
-                // stop if the order_amount is filled.
-                if order_amount.is_zero() {
-                    if !matching_amount.is_zero() {
-                        // todo: push back seems like it should be put to the front?
-                        // Put the last popped out order back in the queue.
-                        orders_queue.push_back(matching_order);
+                    for ask in asks {
+                        /* no self-trading allowed */
+                        if ask.user == order.user {
+                            continue;
+                        }
+
+                        /* determine how much to match */
+                        let amount: U256 = match ask.amount.cmp(&running_total)
+                        {
+                            Ordering::Greater => order.amount,
```

Comment:
Just checking this should be order.amount and not running total? It seems that order.amount gets decreased in the fill function called below but just want to sanity check this

Reply:
@raymogg Good catch; consult 088893a.